Remove unnecessary whitespace in complex tensors

choidongyeon opened the pull request:
This PR addresses Issue #36279. 
Previously, printing of complex tensors would sometimes yield extra spaces before the elements as shown below:
```
print(torch.tensor([[1 + 1.340j, 3 + 4j], [1.2 + 1.340j, 6.5 + 7j]], dtype=torch.complex64))
```
would yield
```
tensor([[(1.0000 + 1.3400j),
         (3.0000 + 4.0000j)],
        [(1.2000 + 1.3400j),
         (6.5000 + 7.0000j)]], dtype=torch.complex64)
```
This occurs primarily because when the max width for the element is being assigned, the formatter's max_width is calculated prior to truncating the float values. As a result, ```self.max_width``` would end up being much longer than the final length of the element string to be printed.

I address this by adding a boolean variable that checks if a complex tensor contains only ints and change the control flow for calculating ```self.max_width``` accordingly.

Here are some sample outputs of both float and complex tensors:

```
tensor([[0., 0.],
        [0., 0.]], dtype=torch.float64) 

tensor([[(0.+0.j), (0.+0.j)],
        [(0.+0.j), (0.+0.j)]], dtype=torch.complex64) 

tensor([1.2000, 1.3400], dtype=torch.float64) 

tensor([(1.2000+1.3400j)], dtype=torch.complex64) 

tensor([[(1.0000+1.3400j), (3.0000+4.0000j)],
        [(1.2000+1.3400j), (6.5000+7.0000j)]], dtype=torch.complex64) 

tensor([1.0000, 2.0000, 3.0000, 4.5000]) 

tensor([(1.+2.j)], dtype=torch.complex64) 
```

cc @ezyang @anjali411 @dylanbespalko

Commit:
Remove unnecessary whitespace in complex tensors

Commit:
Fix lint error

Commit:
Fix lint error

Commit:
Fix lint error

anjali411 commented:
I just saw that this [PR](https://github.com/pytorch/pytorch/pull/35841/files)  leads to this:
```
>>> torch.tensor([1+2j],dtype=torch.complex64)
tensor([(1 + 2.j)], dtype=torch.complex64)
``` 
It should instead output `tensor([(1. + 2.j)], dtype=torch.complex64)` So I think we should also fix this in this PR

anjali411 commented:
also we should move this check inside format function within complex branch and not pass `has_non_zero_decimal_val` in `format(...)` since `has_non_zero_decimal_val` is only used by complex

choidongyeon commented:
@anjali411 Thanks for pointing these out. Will work on them later today.

Commit:
Add missing . in complex tensor print

choidongyeon commented:
> I just saw that this [PR](https://github.com/pytorch/pytorch/pull/35841/files) leads to this:
> 
> ```
> >>> torch.tensor([1+2j],dtype=torch.complex64)
> tensor([(1 + 2.j)], dtype=torch.complex64)
> ```
> 
> It should instead output `tensor([(1. + 2.j)], dtype=torch.complex64)` So I think we should also fix this in this PR

Addressed this in most recent commit. Will fix the second one in a bit.

Commit:
Remove complex tensor decimal check from format

choidongyeon commented:
> also we should move this check inside format function within complex branch and not pass `has_non_zero_decimal_val` in `format(...)` since `has_non_zero_decimal_val` is only used by complex

This was a great suggestion. The only problem is that ```format``` works with single element of a tensor rather than all elements of a tensor. Since we need to check the whole tensor, it wouldn't be possible to put the check inside of ```format```.  However, I was able to work with just the additions I had already made in the ```Formatter``` constructor for this PR. Makes the code much more DRY. Not sure how you feel about the attribute name ```complex_with_decimal``` though?

anjali411 commented on torch/_tensor_str.py:
```diff
@@ -80,8 +81,18 @@ def __init__(self, tensor):
+                                                for value in tensor_view])
             for value in tensor_view:
-                value_str = '{}'.format(value)
+                if self.complex_dtype and self.complex_with_decimal:
```
```
if self.complex_dtype 
   if self.complex_with_decimal:
        value_str = ('{{:.{}f}}').format(PRINT_OPTS.precision).format(value)
   else:
        value_str = "{:.0f}".format(value.item())
```

choidongyeon replied:
attribute name is okay though?

anjali411 replied:
I think we should change it to [this](https://github.com/pytorch/pytorch/pull/36331#discussion_r406515100) :D

anjali411 commented on torch/_tensor_str.py:
```diff
@@ -80,8 +81,18 @@ def __init__(self, tensor):
             tensor_view = tensor.reshape(-1)
 
         if not self.floating_dtype:
+            self.complex_with_decimal = False
```
I think we should change it to `is_non_zero_decimal_val` and perhaps add a comment that it's only used for complex

Commit:
Change attribute name to has_non_zero_decimal_val

Commit:
Remove stray comment

anjali411 commented:
@choidongyeon looks good overall. 
nit-
```
>>> import torch
torch>>> torch.tensor([1+2j])
tensor([(1. + 2.j)], dtype=torch.complex64)
>>> import numpy as np
>>> np.array([1+2j])
array([1.+2.j])
```
we should follow numpy and remove the extra space between real and imag values

Commit:
Remove space between real and imaginary parts of complex number

choidongyeon commented:
> @choidongyeon looks good overall.
> nit-
> 
> ```
> >>> import torch
> >>> torch.tensor([1+2j])
> tensor([(1. + 2.j)], dtype=torch.complex64)
> >>> import numpy as np
> >>> np.array([1+2j])
> array([1.+2.j])
> ```
> 
> we should follow numpy and remove the extra space between real and imag values

Easy peasy. Updated the PR, also updated the PR summary with some sample current outputs.

anjali411 commented:
there's an extra comma that we should get rid of :D

```
>>> torch.randn(2,3,dtype=torch.complex64)
tensor([[(0.3092-1.4706j), (1.0321-0.0074j), (-0.8552-0.3264j)],
        [(-0.6610-0.4281j), (0.4290+0.3676j), (-0.0894+0.2828j)]],
```

anjali411 reviewed (approved):
great job! thanks for working on this :D

choidongyeon commented:
@anjali411 Thanks for being so responsive!
Also, was there originally a comment about an unnecessary comma or am I making this up? I see an email but I don't see it in this thread..

anjali411 commented:
> @anjali411 Thanks for being so responsive!
> Also, was there originally a comment about an unnecessary comma or am I making this up? I see an email but I don't see it in this thread..

of course! let me know if you'd like to work on other issues related to complex numbers. 

yeah I realized it was something else and so I removed it
